Exemplo de PCA

In [7]:
from sklearn.decomposition import PCA
from sklearn.preprocessing import StandardScaler
from sklearn.datasets import load_wine
import pandas as pd

wine = load_wine()

# Converter para DataFrame do Pandas
df_wine = pd.DataFrame(wine.data, columns=wine.feature_names)

# Adicionar a coluna de rótulo (classe do vinho)
df_wine["target"] = wine.target

df_wine.head()

alcohol  malic_acid   ash  alcalinity_of_ash  magnesium  total_phenols  \
0    14.23        1.71  2.43               15.6      127.0           2.80   
1    13.20        1.78  2.14               11.2      100.0           2.65   
2    13.16        2.36  2.67               18.6      101.0           2.80   
3    14.37        1.95  2.50               16.8      113.0           3.85   
4    13.24        2.59  2.87               21.0      118.0           2.80   

   flavanoids  nonflavanoid_phenols  proanthocyanins  color_intensity   hue  \
0        3.06                  0.28             2.29             5.64  1.04   
1        2.76                  0.26             1.28             4.38  1.05   
2        3.24                  0.30             2.81             5.68  1.03   
3        3.49                  0.24             2.18             7.80  0.86   
4        2.69                  0.39             1.82             4.32  1.04   

   od280/od315_of_diluted_wines  proline  target  
0                          3.92   1065.0       0  
1                          3.40   1050.0       0  
2                          3.17   1185.0       0  
3                          3.45   1480.0       0  
4                          2.93    735.0       0

In [8]:
from sklearn.decomposition import PCA
from sklearn.preprocessing import StandardScaler
from sklearn.datasets import load_wine
import pandas as pd

samples = load_wine()
target = samples.target
samples = samples.data
scaler = StandardScaler()
scaled_samples = scaler.fit_transform(samples)


# Create a PCA model with  components: pca
pca = PCA(n_components=2)

# Fit the PCA instance to the scaled samples
pca.fit(scaled_samples)

# Transform the scaled samples: pca_features
pca_features = pca.transform(scaled_samples)

# Print the shape of pca_features
#print(pca_features)
pca_features.shape

(178, 2)

In [9]:
from sklearn.neighbors import KNeighborsClassifier
from sklearn.model_selection import train_test_split
from sklearn.metrics import classification_report

knn = KNeighborsClassifier(n_neighbors=6)

X_train, X_test, y_train, y_test = train_test_split(pca_features, target, test_size=0.3, random_state=42)

knn.fit(X_train, y_train)

y_pred = knn.predict(X_test)

print(classification_report(y_test, y_pred))

              precision    recall  f1-score   support

           0       1.00      0.95      0.97        19
           1       0.95      1.00      0.98        21
           2       1.00      1.00      1.00        14

    accuracy                           0.98        54
   macro avg       0.98      0.98      0.98        54
weighted avg       0.98      0.98      0.98        54



Exemplo de APRIORI

In [15]:
!pip install mlxtend

In [16]:
from mlxtend.frequent_patterns import apriori, association_rules

In [18]:
# Criando um DataFrame de exemplo (1 = produto comprado, 0 = produto não comprado)
data = {
    'Leite': [1, 1, 0, 1, 0, 1],
    'Pão': [1, 1, 1, 1, 0, 1],
    'Café': [0, 1, 1, 0, 1, 1],
    'Ovos': [1, 0, 0, 1, 1, 0],
    'Manteiga': [0, 1, 1, 1, 1, 1]
}

df = pd.DataFrame(data)

# Exibir o dataset de transações
df.head()


/usr/local/lib/python3.11/dist-packages/ipykernel/ipkernel.py:283: DeprecationWarning: `should_run_async` will not call `transform_cell` automatically in the future. Please pass the result to `transformed_cell` argument and any exception that happen during thetransform in `preprocessing_exc_tuple` in IPython 7.17 and above.
  and should_run_async(code)


Leite  Pão  Café  Ovos  Manteiga
0      1    1     0     1         0
1      1    1     1     0         1
2      0    1     1     0         1
3      1    1     0     1         1
4      0    0     1     1         1

In [20]:
# Aplicar Apriori para encontrar os conjuntos de itens frequentes
frequent_itemsets = apriori(df, min_support=0.4, use_colnames=True)

# Exibir os conjuntos frequentes
frequent_itemsets


/usr/local/lib/python3.11/dist-packages/ipykernel/ipkernel.py:283: DeprecationWarning: `should_run_async` will not call `transform_cell` automatically in the future. Please pass the result to `transformed_cell` argument and any exception that happen during thetransform in `preprocessing_exc_tuple` in IPython 7.17 and above.
  and should_run_async(code)
/usr/local/lib/python3.11/dist-packages/mlxtend/frequent_patterns/fpcommon.py:161: DeprecationWarning: DataFrames with non-bool types result in worse computationalperformance and their support might be discontinued in the future.Please use a DataFrame with bool type
  warnings.warn(


support                itemsets
0   0.666667                 (Leite)
1   0.833333                   (Pão)
2   0.666667                  (Café)
3   0.500000                  (Ovos)
4   0.833333              (Manteiga)
5   0.666667            (Leite, Pão)
6   0.500000       (Leite, Manteiga)
7   0.500000             (Pão, Café)
8   0.666667         (Pão, Manteiga)
9   0.666667        (Manteiga, Café)
10  0.500000  (Leite, Pão, Manteiga)
11  0.500000   (Manteiga, Pão, Café)

In [22]:
# Gerar regras de associação
rules = association_rules(frequent_itemsets, metric="lift", min_threshold=1.0)

# Exibir as regras geradas
rules

/usr/local/lib/python3.11/dist-packages/ipykernel/ipkernel.py:283: DeprecationWarning: `should_run_async` will not call `transform_cell` automatically in the future. Please pass the result to `transformed_cell` argument and any exception that happen during thetransform in `preprocessing_exc_tuple` in IPython 7.17 and above.
  and should_run_async(code)


antecedents        consequents  antecedent support  \
0             (Leite)              (Pão)            0.666667   
1               (Pão)            (Leite)            0.833333   
2          (Manteiga)             (Café)            0.833333   
3              (Café)         (Manteiga)            0.666667   
4   (Leite, Manteiga)              (Pão)            0.500000   
5     (Pão, Manteiga)            (Leite)            0.666667   
6             (Leite)    (Pão, Manteiga)            0.666667   
7               (Pão)  (Leite, Manteiga)            0.833333   
8     (Pão, Manteiga)             (Café)            0.666667   
9         (Pão, Café)         (Manteiga)            0.500000   
10         (Manteiga)        (Pão, Café)            0.833333   
11             (Café)    (Pão, Manteiga)            0.666667   

    consequent support   support  confidence   lift  representativity  \
0             0.833333  0.666667        1.00  1.200               1.0   
1             0.666667  0.666667        0.80  1.200               1.0   
2             0.666667  0.666667        0.80  1.200               1.0   
3             0.833333  0.666667        1.00  1.200               1.0   
4             0.833333  0.500000        1.00  1.200               1.0   
5             0.666667  0.500000        0.75  1.125               1.0   
6             0.666667  0.500000        0.75  1.125               1.0   
7             0.500000  0.500000        0.60  1.200               1.0   
8             0.666667  0.500000        0.75  1.125               1.0   
9             0.833333  0.500000        1.00  1.200               1.0   
10            0.500000  0.500000        0.60  1.200               1.0   
11            0.666667  0.500000        0.75  1.125               1.0   

    leverage  conviction  zhangs_metric  jaccard  certainty  kulczynski  
0   0.111111         inf       0.500000      0.8       1.00        0.90  
1   0.111111    1.666667       1.000000      0.8       0.40        0.90  
2   0.111111    1.666667       1.000000      0.8       0.40        0.90  
3   0.111111         inf       0.500000      0.8       1.00        0.90  
4   0.083333         inf       0.333333      0.6       1.00        0.80  
5   0.055556    1.333333       0.333333      0.6       0.25        0.75  
6   0.055556    1.333333       0.333333      0.6       0.25        0.75  
7   0.083333    1.250000       1.000000      0.6       0.20        0.80  
8   0.055556    1.333333       0.333333      0.6       0.25        0.75  
9   0.083333         inf       0.333333      0.6       1.00        0.80  
10  0.083333    1.250000       1.000000      0.6       0.20        0.80  
11  0.055556    1.333333       0.333333      0.6       0.25        0.75

- Quando compram leite tem 83% de chance de comprar pão
- Quando compram café tem 83% de chance de comprar manteiga